File: ios/Sources/UI/TableRow.swift
import RealityKit
import SmashCore

/// One row of a table — a league table line, a list entry. A slab with cells of lettering laid
/// out by column, an optional kit chip, and a slot it springs to: re-rank the table and the rows
/// hop past each other into their new places.
@MainActor
final class TableRow: Semantic, Presentable {
    struct Column {
        /// The column's centre, as a share of the row width from its left edge (0…1).
        let at: Float
        let align: Label3D.Align
    }

    struct Kit { let primary: Int; let secondary: Int }

    let entity = Entity()
    private let body = Entity()
    private let slab: ModelEntity
    private var cells: [ModelEntity] = []
    private var cellNodes: [Entity] = []
    private let columns: [Column]
    let size: SIMD2<Float>
    var rest = Transform()
    var presence: Presence
    private(set) var semantics: Semantics
    private var slot: Spring
    private var hop: Jiggle
    private let textHeight: Float
    private let motion: MotionTokens

    init(_ texts: [String], columns: [Column], size: SIMD2<Float>, id: String,
         colour: Int, ink: Int = DesignTokens.Colour.ink, kit: Kit? = nil, kitAt: Float = 0.13,
         textHeight: Float = DesignTokens.Size.textBody, y: Float = 0,
         entrance: Entrance = .slide(fromLeft: true), motion: MotionTokens) {
        precondition(texts.count == columns.count, "one text per column")
        self.columns = columns
        self.size = size
        self.textHeight = textHeight
        self.motion = motion
        presence = Presence(entrance, motion: motion)
        slot = Spring(motion.spring(.pop), initial: Double(y))
        hop = Jiggle(motion.spring(.wobbly))
        semantics = Semantics(id: id, label: texts.joined(separator: ", "), trait: .staticText)
        let depth = DesignTokens.Size.slabDepth * 0.6
        slab = Blocks.slab([size.x, size.y, depth], colour, corner: size.y * 0.3)
        entity.addChild(body)
        body.addChild(slab)
        for (text, col) in zip(texts, columns) {
            let m = Blocks.text(text, height: textHeight, ink)
            let node = Entity()
            node.addChild(m)
            node.position.z = depth / 2
            body.addChild(node)
            cells.append(m)
            cellNodes.append(node)
            place(node, m, col)
        }
        if let kit {
            let chip = Entity()
            let shirt = Blocks.slab([size.y * 0.62, size.y * 0.62, depth * 0.8], kit.primary, corner: size.y * 0.12)
            let stripe = Blocks.slab([size.y * 0.2, size.y * 0.64, depth * 0.84], kit.secondary, corner: 0.005)
            chip.addChild(shirt)
            chip.addChild(stripe)
            chip.position = [-size.x / 2 + kitAt * size.x, 0, depth / 2]
            body.addChild(chip)
        }
        rest.translation.y = y
        entity.isEnabled = false
    }

    private func place(_ node: Entity, _ m: ModelEntity, _ col: Column) {
        let cx = -size.x / 2 + col.at * size.x
        let w = Blocks.width(of: m)
        switch col.align {
        case .centre: node.position.x = cx
        case .leading: node.position.x = cx + w / 2
        case .trailing: node.position.x = cx - w / 2
        }
    }

    var boundsEntity: Entity { entity }
    var bounds: BoundingBox {
        BoundingBox(min: [-size.x / 2, -size.y / 2, -0.03], max: [size.x / 2, size.y / 2, 0.03])
    }
    var isPresent: Bool { presence.isSettledIn }

    func show(after delay: Double) { presence.show(after: delay) }
    func hide(after delay: Double) { presence.hide(after: delay) }

    func set(_ texts: [String]) {
        for i in cells.indices where i < texts.count {
            Blocks.retext(cells[i], texts[i], height: textHeight)
            place(cellNodes[i], cells[i], columns[i])
        }
        semantics.label = texts.joined(separator: ", ")
    }

    func recolour(_ rgb: Int) { Blocks.recolour(slab, rgb) }

    /// Springs to a new vertical slot, with a hop if it moved.
    func move(toY y: Float) {
        guard abs(Double(y) - slot.target) > 0.001 else { return }
        let up = Double(y) > slot.target
        slot.target = Double(y)
        hop.kick(twist: (up ? 1 : -1) * motion.kick(.celebrate) * 0.25, swell: motion.kick(.celebrate) * 0.2)
    }

    func update(_ dt: Double, _ ctx: UIContext) {
        if ctx.reduceMotion { slot.snap(to: slot.target) } else { slot.advance(dt) }
        rest.translation.y = Float(slot.value)
        // A row climbing the table comes forward to pass over the others; a falling one ducks back.
        rest.translation.z = Float(max(-0.08, min(0.12, (slot.target - slot.value) * 0.5)))
        presence.advance(dt, ctx)
        presence.apply(to: entity, rest: rest, reduceMotion: ctx.reduceMotion)
        guard entity.isEnabled else { return }
        hop.advance(dt, reduceMotion: ctx.reduceMotion)
        body.orientation = hop.rotation
        body.scale = SIMD3(repeating: hop.scale)
    }
}

File: android/app/src/main/kotlin/in/nann/smashhockey/ui/TableRow.kt
package `in`.nann.smashhockey.ui

import `in`.nann.smashhockey.ui.generated.DesignTokens.Colour
import `in`.nann.smashhockey.ui.generated.DesignTokens.Size
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import `in`.nann.smashhockey.core.feel.TextLayout
import `in`.nann.smashhockey.engine.Spring

/**
 * One row of a table — a league table line, a list entry — the twin of iOS's `TableRow`. A slab
 * with cells of lettering laid out by column, an optional kit chip, and a slot it springs to:
 * re-rank the table and the rows hop past each other into their new places.
 */
class TableRow(
    private val kit: Kit,
    texts: List<String>,
    private val columns: List<Column>,
    val w: Float, val h: Float,
    id: String,
    colour: Int,
    ink: Int = Colour.INK,
    chip: KitColours? = null,
    kitAt: Float = 0.13f,
    private val textHeight: Float = Size.TEXT_BODY,
    y: Float = 0f,
    entrance: Entrance = Entrance.Slide(fromLeft = true),
    /** Said after the row's cells, so TalkBack names what opening it gives. */
    private val hint: String? = null,
    /** What a tap on the row does (§16.3a); null leaves it lettering that takes no fingers. */
    private val action: (() -> Unit)? = null,
) : Interactive, Presentable {
    /** A column's centre, as a share of the row width from its left edge (0…1). */
    class Column(val at: Float, val align: Label3D.Align)

    class KitColours(val primary: Int, val secondary: Int)

    companion object {
        private fun spoken(texts: List<String>, hint: String?) =
            (if (hint == null) texts else texts + hint).joinToString(", ")
    }

    override val node = kit.node(null)
    private val body = kit.node(node)
    private val slab: UiNode
    private val cells = ArrayList<UiNode>()
    private val cellNodes = ArrayList<UiNode>()
    override val rest = Xform()
    val presence = Presence(entrance, kit.motion)
    override val semantics = Semantics(
        id, spoken(texts, hint),
        trait = if (action == null) Semantics.Trait.STATIC_TEXT else Semantics.Trait.BUTTON,
    )
    private var held = false
    private val slot = Spring(kit.motion.spring(SpringName.POP), y.toDouble())
    private val hop = Jiggle(kit.motion.spring(SpringName.WOBBLY))
    private val motion = kit.motion

    init {
        require(texts.size == columns.size) { "one text per column" }
        val depth = Size.SLAB_DEPTH * 0.6f
        slab = kit.slab(w, h, depth, colour, body, corner = h * 0.3f)
        for ((text, col) in texts.zip(columns)) {
            val cell = kit.node(body)
            val m = kit.text(text, textHeight, ink, cell)
            cell.transform.tz = depth / 2
            cells += m
            cellNodes += cell
            place(cell, m, col)
        }
        if (chip != null) {
            val c = kit.node(body)
            kit.slab(h * 0.62f, h * 0.62f, depth * 0.8f, chip.primary, c, corner = h * 0.12f)
            kit.slab(h * 0.2f, h * 0.64f, depth * 0.84f, chip.secondary, c, corner = 0.005f)
            c.setPosition(-w / 2 + kitAt * w, 0f, depth / 2)
        }
        rest.ty = y
        node.enabled = false
    }

    private fun place(cell: UiNode, m: UiNode, col: Column) {
        val cx = -w / 2 + col.at * w
        cell.transform.tx = cx + TextLayout.alignX(col.align.shared, kit.width(m).toDouble()).toFloat()
        cell.changed()
    }

    override val boundsNode get() = node
    override val bounds = Bounds(-w / 2, -h / 2, -0.03f, w / 2, h / 2, 0.03f)
    override val isPresent get() = presence.isSettledIn

    override fun show(after: Double) = presence.show(after)
    override fun hide(after: Double) {
        held = false
        presence.hide(after)
    }

    fun set(texts: List<String>) {
        for (i in cells.indices) {
            if (i >= texts.size) break
            kit.retext(cells[i], texts[i], textHeight)
            place(cellNodes[i], cells[i], columns[i])
        }
        semantics.label = spoken(texts, hint)
    }

    fun recolour(rgb: Int) = slab.recolour(rgb)

    // ------ taps (§16.3a): a row with something to do sinks under the finger and opens on the lift

    override val takesTouches get() = action != null

    override fun touchDown(ray: TouchRay) {
        if (action == null) return
        held = true
        hop.kick(0.0, -motion.kick(KickName.CELEBRATE) * 0.2)
        KitSound.press()
    }

    override fun touchUp(ray: TouchRay, inside: Boolean) {
        if (!held) return
        held = false
        if (inside) action?.invoke()
    }

    override fun activate() {
        val a = action ?: return
        hop.kick(0.0, motion.kick(KickName.CELEBRATE) * 0.2)
        KitSound.press()
        a()
    }

    /** Springs to a new vertical slot, with a hop if it moved. */
    fun move(toY: Float) {
        if (abs(toY - slot.target) <= 0.001) return
        val up = toY > slot.target
        slot.target = toY.toDouble()
        val k = motion.kick(KickName.CELEBRATE)
        hop.kick((if (up) 1 else -1) * k * 0.25, k * 0.2)
    }

    override fun update(dt: Double, ctx: UiContext) {
        if (ctx.reduceMotion) slot.snap(slot.target) else slot.advance(dt)
        rest.ty = slot.value.toFloat()
        // A row climbing the table comes forward to pass over the others; a falling one ducks back.
        rest.tz = max(-0.08, min(0.12, (slot.target - slot.value) * 0.5)).toFloat()
        presence.advance(dt, ctx)
        presence.apply(node, rest, ctx.reduceMotion)
        if (!node.enabled) return
        hop.advance(dt, ctx.reduceMotion)
        body.transform.rot.set(hop.rotation)
        body.setScale(hop.scale)
    }
}
